### PR TITLE
Fine-tuning and fixes in import endpoint

### PR DESCRIPTION
**Pull request by michaljurecko**
Changes:
- Target `SliceKey` for a new record is obtained correctly (not implemented before).
- Clearer and simpler Jsonnet functions.
- Code readability increased and additional tests added.
- Optimized speed.
- Complex code has been divided into several smaller packages/files.
- Fixed stats sync (fixed too many keys in the transaction error).

-----

Vyriesil som viacero edge-cases a problemov, na ktore som narazil pri testovani.

Asi o `10%` sa zvysila aj rychlost:

`static.js` benchmark:
image

`template.js` benchmark:
image

### CODE REVIEW

File: internal/pkg/service/buffer/api/service/import.go [hunk 1]
```diff
@@ -38,90 +34,43 @@ func (s *service) Import(d dependencies.ForPublicRequest, payload *buffer.Import
...
-	return errs.ErrorOrNil()
-}
-
-func parseRequestBody(contentType string, reader io.ReadCloser) (res *orderedmap.OrderedMap, err error) {
-	if !isContentTypeForm(contentType) && !regexp.MustCompile(`^application/([a-zA-Z0-9\.\-]+\+)?json$`).MatchString(contentType) {
-		return nil, NewUnsupportedMediaTypeError(errors.New(
-			"Supported media types are application/json and application/x-www-form-urlencoded.",
-		))
+		// Update statistics
+		s.stats.Notify(sliceKey, uint64(len(csvRow)))
```
Comment (michaljurecko) [inline internal/pkg/service/buffer/api/service/import.go:69]: Uz sa spravne reportuje do statistik sirka celeho riadku, predtym sa to pocitalo trocha zvlastne po stlpcoch.

File: internal/pkg/service/buffer/store/receiver.go [hunk 2]
```diff
@@ -156,6 +156,7 @@ func (s *Store) DeleteReceiver(ctx context.Context, receiverKey key.ReceiverKey)
 		s.deleteReceiverExportsOp(ctx, receiverKey),
 		s.deleteReceiverMappingsOp(ctx, receiverKey),
 		s.deleteReceiverTokensOp(ctx, receiverKey),
+		s.deleteReceiverStatsOp(ctx, receiverKey),
```
Comment (michaljurecko) [inline internal/pkg/service/buffer/store/receiver.go:159]: Zmazanie statistik pri zmazani receiver.

File: internal/pkg/service/buffer/store/stats.go [hunk 3]
```diff
@@ -2,22 +2,52 @@ package store
...
 	"github.com/keboola/keboola-as-code/internal/pkg/service/buffer/store/model"
 	"github.com/keboola/keboola-as-code/internal/pkg/service/common/etcdop/op"
 	"github.com/keboola/keboola-as-code/internal/pkg/telemetry"
+	"github.com/keboola/keboola-as-code/internal/pkg/utils/errors"
 )
 
+const maxStatsPerTxn = 50
+
 func (s *Store) UpdateSliceStats(ctx context.Context, nodeID string, stats []model.SliceStats) (err error) {
-	_, span := s.tracer.Start(ctx, "keboola.go.buffer.store.UpdateStats")
```
Comment (michaljurecko) [inline internal/pkg/service/buffer/store/stats.go:12]: Naraz aktualizujeme statistiky pre viacere slices v jednej transakcii.
Transakcia moze mat ale maximalne `128` poloziek, preto sa to musi podelit (po `50`).

File: test/api/buffer/import/import/005-400-empty-payload/expected-response.json [hunk 4]
```diff
@@ -1,5 +1,5 @@
 {
   "statusCode": 400,
   "error": "buffer.badRequest",
-  "message": "Empty request body."
+  "message": "Failed to format record for export \"export-1\":\n- path \"col1\" not found in the body."
```
Comment (michaljurecko) [inline test/api/buffer/import/import/005-400-empty-payload/expected-response.json:4]: Empty JSON je validny payload, ale chyba tam hladany key.

File: test/api/buffer/import/import/expected-etcd-kvs.txt [hunk 5]
```diff
@@ -128,7 +126,7 @@
 <<<<<
 /record/%%TEST_KBC_PROJECT_ID%%/receiver-1/export-1/%sZ/%sZ_%c%c%c%c%c
 -----
-<<~~id~~>>,"{""col1"":""val3"",""col2"":""val4""}","{""Accept-Encoding"":""gzip"",""Content-Length"":""19"",""Content-Type"":""application/x-www-form-urlencoded"",""User-Agent"":""%s""}",%s,%s,"""val3:val4"""
+<<~~id~~>>,col1=val3&col2=val4,"{""Accept-Encoding"":""gzip"",""Content-Length"":""19"",""Content-Type"":""application/x-www-form-urlencoded"",""User-Agent"":""%s""}",%s,%s,"""val3:val4"""
```
Comment (michaljurecko) [inline test/api/buffer/import/import/expected-etcd-kvs.txt:129]: Body column type uklada raw body, teda vieme prijat aj ine typy requestov, ako som uz pisal.
(Rovnako ako `BodyStr()` v sablone, ... no `Body(...)` v sablone vrati uz objekt, request sa vtedy parsuje.)

File: internal/pkg/service/buffer/api/service/import.go [hunk 6]
```diff
@@ -38,90 +34,43 @@ func (s *service) Import(d dependencies.ForPublicRequest, payload *buffer.Import
...
-			errs.AppendWithPrefixf(err, `failed to create record for export "%s"`, e.ExportID)
+			errs.Append(err)
+			continue
 		}
 
-		size := uint64(0)
-		for _, column := range csv {
-			size += uint64(len(column))
+		// Persist record
+		sliceKey := export.OpenedSlice.SliceKey
```
Comment (michaljurecko) [inline internal/pkg/service/buffer/api/service/import.go:61]: Uz sa pouziva realny `SliceKey`.

File: test/api/buffer/import/import/expected-etcd-kvs.txt [hunk 7]
```diff
@@ -186,23 +184,8 @@
...
-{
-  "projectId": %%TEST_KBC_PROJECT_ID%%,
-  "receiverId": "receiver-1",
-  "exportId": "export-1",
-  "fileId": "%s",
-  "sliceId": "%s",
-  "count": 1,
-  "size": %d,
+  "count": 2,
+  "size": 567,
```
Comment (michaljurecko) [inline test/api/buffer/import/import/expected-etcd-kvs.txt]: Pouziva sa uz spravny `SliceKey`, takze tu nemame 2 rozne statistiky, ale jednu (kedze predtym bolo `SliceKey = now`).

File: test/api/buffer/import/import/expected-etcd-kvs.txt [hunk 8]
```diff
@@ -185,7 +185,7 @@
   "fileId": "%s",
   "sliceId": "%s",
   "count": 2,
-  "size": 567,
+  "size": %d,
```
Comment (michaljurecko) [inline test/api/buffer/import/import/expected-etcd-kvs.txt:188]: Z nejakeho dovodu dostavam lokalne IP adresu `127.0.0.1`, a v CI `::1` (IPv6), teda `size` moze mat roznu hodnotu.

File: internal/pkg/service/buffer/api/receive/jsonnet/jsonnet.go [hunk 9]
```diff
@@ -31,7 +31,7 @@ const (
 	// DefaultTimeFormat with fixed length, so it can be used for lexicographic sorting in the target database.
 	// Value is RFC3339 and ISO8601 compatible.
 	DefaultTimeFormat   = "%Y-%m-%dT%H:%M:%S.%fZ"
-	ThrowErrOnUndefined = "<<~~errOrUndefined~~>>"
+	ThrowErrOnUndefined = "<<~~errOnUndefined~~>>"
```
Comment (michaljurecko) [inline internal/pkg/service/buffer/api/receive/jsonnet/jsonnet.go:34]: Typo.

**Review by michaljurecko — COMMENTED** [review]
Okomentovane po commitoch.

File: test/api/buffer/import/import/expected-etcd-kvs.txt (hunk 8 above)
Comment (JakubMatejka) [reply test/api/buffer/import/import/expected-etcd-kvs.txt:188]: Na to už jsem taky při něčem nedávno narazil. 🙂

**Review by JakubMatejka — APPROVED** [review]
